### DISCUSSION TITLE
📂 A Synthetic X11 Drag&drop "protocol" was Implemented, but...

### DISCUSSION THREAD

**Discussion by GiorgosXou** [GiorgosXou/TUIFIManager#92] [Announcements]
# Intro
**In an attempt to be as much inclusive** as of the fact: that this is going to be used across multiple distridutions of linux, and for the sake of *(my sanity and)* future portability across different OS, terminals and display servers such as wayland *(by users who might want to do so)*, **I took a slightly unorthodox aprroach to the problem... a simple, intuitive but incomplete one!** *([Addressing Issue 21](https://github.com/GiorgosXou/TUIFIManager/issues/21))*

# Approach 
**I'll start by saying:** there's a good reason why [dragon](https://github.com/mwh/dragon) exists; it's not as easy as it sounds to syntheticly\\virtually perform drag and drop *(especially when you try to be as much inclusive as possible [...])*. **After alot of pitfalls** and some deep research, **I literally just ended up** using a `pyside6`-gui that appears the momment you are about to drop something in your *(script-running)*-terminal + for the TUI-to-GUI operation: a fake `QDrag` signal-event situation as seen and demonstrated bellow.
 
![dnd_2_Peek 2024-03-30 17-08 (2)](https://github.com/GiorgosXou/TUIFIManager/assets/11987271/408d7342-ef0c-4dce-92d7-77f9a0e082d2)

***!!!! Please consider trying it and telling me if it works for you, I need your feedback! Code is here bellow:***
```
pip install PySide6 python-xlib pynput
```

<details>
  <summary>👉👉👉👉 Click to expand code</summary>

```python
from pynput.mouse import Button, Controller, Listener
from PySide6.QtWidgets import QApplication, QMainWindow
from PySide6.QtCore import Qt, QMimeData, QUrl, QObject, Signal
from PySide6.QtGui import QDragEnterEvent, QDropEvent, QDrag
from Xlib import display
from time import sleep


TEMP_FILE = "home/xou/Downloads/me.png"


class DragDropWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowFlag(Qt.X11BypassWindowManagerHint, True)
        self.setWindowFlag(Qt.WindowStaysOnTopHint, True)
        self.setWindowFlag(Qt.FramelessWindowHint, True)
        self.setWindowOpacity(0) 
        self.setAcceptDrops(True)
        self.resize(9,9)
        # self.setStyleSheet("background-color: rgba(0, 0, 0, 0);")  # 128 is the alpha value for 50% opacity
        
    def dragEnterEvent(self, event: QDragEnterEvent):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
        else:
            event.ignore()

    def dropEvent(self, event: QDropEvent):
        a = event.mimeData().text()
        if a:
            print("Dropped  URL: ", event.mimeData().text())
        else:
            for url in event.mimeData().urls():
                file_path = url.toLocalFile()
                print("Dropped file:", file_path)



d           = display.Display()
root        = d.screen().root
self_window = d.get_input_focus().focus # or parrent but if not parrent is root (because some windows managers such i3 use containers\windows)
listener    = Listener()
app         = QApplication()
vdnd_win    = DragDropWindow() # virtual dnd win
mouse       = Controller()



def is_mouse_cursor_inside(geometry, x,y):
    return (x > geometry.x and x < geometry.x + geometry.width and y > geometry.y and y < geometry.y + geometry.height)

def get_self_geometry():
    if self_window.query_tree().parent != root: # at least for i3 (because some windows managers such i3 use plus containers\windows)
         return self_window.query_tree().parent.get_geometry()
    else: # I tried to be inclusive, even thought I have no idea of how other window managers might handle this
         return self_window.get_geometry()


class MyObject(QObject):
    my_signal = Signal() # Define a custom signal
    def __init__(self): super().__init__()
    def begin_virtial_dnd(self): self.my_signal.emit() # Emit the signal with some data


obj = MyObject()
w = QMainWindow()

def dnd_slot():
    print(f"Dropping: {TEMP_FILE}")
    url = QUrl.fromLocalFile(TEMP_FILE)
    mime_data = QMimeData()
    mime_data.setUrls([url])  # Set your file path here
    drag = QDrag(w)
    drag.setMimeData(mime_data)
    drag.exec()



def perform_virtual_file_dnd(x,y):
    obj.begin_virtial_dnd()  # This will emit the signal
    sleep(.3) # ^^^ debatable... but what can i do? Also SuS...
    # mouse.release(Button.left)
    # sleep(.01)
    mouse.move(1,1)
    sleep(.01)
    mouse.move(1,1)
    sleep(.01)
    mouse.move(-1,-1)
    sleep(.01)
    mouse.move(-1,-1)
    mouse.click(Button.left)



mouse_moved_outside = False
def handle_dnd_tui_to_gui(x, y): # TUI to whatever
    # print(f"TUI -> GUI X:{x} y:{y}")
    if not is_mouse_cursor_inside(get_self_geometry(),x,y):
        global mouse_moved_outside
        mouse_moved_outside = True


def handle_dnd_gui_to_tui(x, y): # Gui to potential TUI
    # print(f"GUI -> TUI X:{x} y:{y}")
    if is_mouse_cursor_inside(get_self_geometry(),x,y) and self_window.get_attributes().map_state == 2: # 2 meaning is visible | prevents issues when you are in another i3 tab
        vdnd_win.move(x-5,y-5)
        vdnd_win.show()
    else:
        vdnd_win.hide()


suppress = False
def on_mouse_click(x, y, button, pressed):
    global suppress
    global listener
    global mouse_moved_outside
    if suppress: 
        suppress = False
        return
    if pressed and button == Button.left:
        if d.get_input_focus().focus == self_window:
            listener.on_move = handle_dnd_tui_to_gui
        else:
            listener.on_move = handle_dnd_gui_to_tui
    else:
        if mouse_moved_outside and listener.on_move == handle_dnd_tui_to_gui:
            mouse_moved_outside = False
            suppress = True
            listener.on_click = lambda *args : None # might change it to AbstractListener
            listener.on_move = lambda *args : None # might change it to AbstractListener
            perform_virtual_file_dnd(x,y)
            listener.on_click = on_mouse_click # this won't work by itself, as it will "backfire" an ivent from perform_virtual_file_dnd, even though we did lambda *args : None 
        listener.on_move = lambda *args : None # might change it to AbstractListener
        vdnd_win.hide()


def main():
    obj.my_signal.connect(dnd_slot)  # Connect the signal to a slot (a function)
    global listener
    listener.on_click=on_mouse_click
    listener.start()
    app.exec()


if __name__ == "__main__":
    main()
```
</details>

## Known Issues
1. It doesn't work as expected with overlapping windows

# Research

- ***Used:***
- - [pynput](https://github.com/moses-palmer/pynput)
- - [python-xlib](https://github.com/python-xlib/python-xlib)
- ***Xorg:***
- - [[LV2] Drag and drop (XDND)](http://lists.lv2plug.in/pipermail/devel-lv2plug.in/2020-August/002004.html)
- - [Xlib - C Language X Interface](https://www.x.org/releases/X11R7.7/doc/libX11/libX11/libX11.html)
- - [X11 Input Extension Protocol Specification](https://www.x.org/docs/Xi/proto.pdf)
- - [Drag-and-Drop Protocol for the X Window System](https://www.freedesktop.org/wiki/Specifications/XDND/)
- - [Xlib difference between _NET_ACTIVE_WINDOW and XGetInputFocus](https://stackoverflow.com/questions/31800880/xlib-difference-between-net-active-window-and-xgetinputfocus)
- - [11.8.2 Using the Default Error Handlers](https://tronche.com/gui/x/xlib/event-handling/protocol-errors/default-handlers.html)
- - [10.13.1 ClientMessage Events](https://tronche.com/gui/x/xlib/events/client-communication/client-message.html)
- - [XQueryTree](https://tronche.com/gui/x/xlib/window-information/XQueryTree.html)
- - ***Python:***
- - - [Obtain Active window using Python](https://stackoverflow.com/questions/10266281/obtain-active-window-using-python)
- - - [list all open (X11) windows in Python in Z-Order](https://stackoverflow.com/questions/52545937/how-can-i-list-all-open-x11-windows-on-gnu-linux-from-a-python-script)
- - ***Projects:*** 
- - - [tkinter/dnd.py](https://github.com/python/cpython/blob/main/Lib/tkinter/dnd.py)
- - - [Xorg-clipboard-xdnd](https://github.com/frank038/Xorg-clipboard-xdnd?tab=readme-ov-file)
- - - [xlib_xdnd](https://github.com/PhilPotter/xlib_xdnd)
- - - [sample-implementation](https://johnlindal.wixsite.com/xdnd/sample-implementation)
- - ***Random:***
- - - [XCB + record x11 extension example](https://stackoverflow.com/questions/20823894/xcb-record-x11-extension-example)
- - - [window.get_wm_name() is empty but not in window.get_property() equivalent](https://github.com/python-xlib/python-xlib/issues/212)
- - - [pyglet - xdnd something](https://github.com/encukou/pyglet/blob/382c6767f4191632c0a8f6c78948c260ab2a5fd1/pyglet/window/xlib/__init__.py#L83)
- - - [Enumerating all visible desktop windows by z-order](https://gist.github.com/blewert/b6e7b11c565cf82e7d700c609f22d023)
- - - [keylogger.py](https://gist.github.com/whym/402801)
- ***Wayland:***
- - [Wayland alternative for xorg's xdotool?](https://superuser.com/questions/1032270/wayland-alternative-for-xorgs-xdotool)
- ***QT:***
- - ***Random:***
- - - https://stackoverflow.com/questions/66842999/create-a-transparent-blur-window
- ***Utilities:***
- - [xscope](https://aur.archlinux.org/packages/xscope)

# Outro
**First of all thanks** for reading so far. Second of all, this is the best I can do without any donation, considering the situation in which I am. So far I had only one donation from a single person,  [to which I am really greatfull](https://github.com/GiorgosXou/TUIFIManager/discussions/71) and wish the best for him or her for giving me those special 8$. So, if you don't want to support me economically, **at least consider** responding to this post by telling me if the code is working for you **or** [upvote this post so it will get more attention](https://stackoverflow.com/questions/76523336/simulate-drag-and-drop-from-terminal-in-x11). I'm programming because I love to. And again thanks... for your support and love ⭐

[![Donate](https://img.shields.io/badge/Donate-PayPal-blue.svg)](https://www.paypal.com/donate/?hosted_button_id=QNQN23M55EJVS)

Monero Address: 897ehhSQJQpGF7tYDhQM51jiX7nnHmzuYAW4q8JGwJxu8JKXvaK6AivCzatuJxnifjZ2qy98ks2g2PhmTaYCMMta2Ga2LJx

**Comment by ilhan-athn7** [comment]
https://github.com/GiorgosXou/TUIFIManager/assets/80551220/aa43ee55-c679-4f58-b55b-d7c227516808

it is working

**Reply by GiorgosXou** [reply]
As expected... commenting out everything under `perform_virtual_file_dnd()` besides the first line *(`obj.begin_virtial_dnd()`)*, should fix the issue of dropping files to GUIs with the only catch that you will need to manually click once more after it.

**Comment by GiorgosXou** [comment]
Additionally, I'd like to notice that it's eay to check which windows are overlaping and ignore them via something like:
```python
def get_windows(window):
    try:
        yield window
        for child_window in window.query_tree().children:
            yield from get_windows(child_window)
    except error.BadWindow as e: # result of query_tree()
        return
...
        
for win in get_windows(root):
    # ...
```
via checking if wins after a condition like `if win == self_window`, are overlaping in Z-Order... The only issue it that Xorg or whatever displays a RECTANGLE/window under cursor that I can't distinguish from other overlapping onces, making it complex

**Comment by GiorgosXou** [comment]
for god's sake! why it has to be so complicated to run a QT app in another thread besides main...
https://github.com/midjji/convenient_multithreaded_qt_gui/issues/1

**Reply by GiorgosXou** [reply]
HA! I found a solution!!

**Reply by ilhan-athn7** [reply]
good luck

**Comment by GiorgosXou** [comment]
![sctg_2024-04-07 16:09:38 907354](https://github.com/GiorgosXou/TUIFIManager/assets/11987271/3fb8ca8a-d0ad-4f6a-b42a-e87c3436b8c6)
![dnd_tuifi_done_Peek 2024-04-07 16-10](https://github.com/GiorgosXou/TUIFIManager/assets/11987271/acc2b273-aac4-44d3-92f2-e866f71aa6e2)

\* dropped
